pycross/private/tools/pdm_translator.py

```diff
@@ -0,0 +1,157 @@
+import argparse
+import os
+import sys
+from pathlib import Path
+
+from packaging.utils import Version
+
+from pycross.private.tools.lock_model import LockSet
+from pycross.private.tools.lock_model import Package
+from pycross.private.tools.lock_model import PackageDependency
+from pycross.private.tools.lock_model import PackageFile
+from pycross.private.tools.lock_model import package_canonical_name
+
+# FIXME dirty hack to "fix" pdm-namespace shadowing ...
+for p in sys.path:
+    if "pdm_pep517/site-packages" in p:
+        os.remove(f"{p}/pdm/__init__.py")
+
+from pdm.core import Core as PDMCore
+from pdm.project import Project as PDMProject
+from pdm.models.candidates import Candidate as PDMCandidate
+from pdm.models.repositories import BaseRepository as PDMBaseRepository
+
+
+class MismatchedVersionException(Exception):
+    pass
+
+
+class Project(PDMProject):
+    def __init__(self, project_file: Path, lock_file: Path) -> None:
+        super().__init__(
+            core=PDMCore(),
+            root_path=project_file.parent.absolute(),
+        )
+        self._lockfile_file = lock_file.absolute()
+        self._pyproject_file = project_file.absolute()
+
+    @property
+    def pyproject_file(self) -> Path:
+        return self._pyproject_file
+
+
+def candidate_package_key(candidate: PDMCandidate) -> str:
+    return f"{candidate.name}@{candidate.version}"
+
+
+def parse_candidate_file(file_name: str, file_hash: str) -> PackageFile:
+    assert file_hash.startswith("sha256:")
+    return PackageFile(name=file_name, sha256=file_hash[7:])
+
+
+def translate(project_file: str, lock_file: str) -> LockSet:
+    project = Project(project_file=Path(project_file), lock_file=Path(lock_file))
+    repository = project.locked_repository
+    pinned_keys = {}
+    lock_packages = {}
+
+    for _dependency in project.dependencies.values():
+        for _candidate in repository.find_candidates(_dependency):
+            # TODO handle the case where multiple candidates are found
+            pin = package_canonical_name(_candidate.name)
+            pinned_keys[pin] = candidate_package_key(_candidate)
+
+    # TODO how do we handle packages with extras? do merge them with the package without extras ?
```

Comment:
I think this extras handling might be one of the reasons I started using Poetry instead. But I don't recall completely. If you're curious, the earliest revisions in this repo use PDM, I think.

Reply:
i agree. I'll change the comments accordingly to document that decision then